Lock dossier subtree during resolve transition

lukasgraf opened the pull request:
Implements a **locking mechanism to prevent concurrent dossier resolution**.

This mechanism is intended to prevent users from simultaneously triggering the `resolve` transition for a dossier (which would result in degraded performance because retries of conflicting transactions).

We do this by issuing a **persistent lock** that gets committed in its own transaction as the very first thing in the resolution process. Further attempts at resolving a dossier are then rejected as long as such a lock exists (and hasn't expired). Once the dossier is resolved successfully, the lock is removed and the removal will be committed again. The lock will **expire after 24h**, and which point it will just be treated as if it wouldn't exist.

In case of an exception, we catch it, abort the transaction, remove the lock, commit the removal, and re-reaise the exception to be handled by the usual error handling in `ZPublisher`.

When checking whether a subdossier is locked, possible locks on any of its parents are also considered. So basically an **entire dossier subtree is locked as a whole**, by attaching a lock to the main dossier. If a dossier is locked, this is displayed by appending a message to the workflow status in the byline - this is however only displayed for the current context, not determined recursively (for performance reasons).

In addition, this change includes a **guard condition to prevent dossiers from being resolved twice**.

Logging is currently very much on the verbose side, and could eventually be toned down once this proves to be a reliable strategy in the field.

**Note:** I added a check for a dirty transaction when committing the lock. When the transaction is found to be dirty (which it usually shouldn't be), this is logged to Sentry, but we still proceed with committing. 

#### View of a dossier during resolve process
![view_during_resolve](https://user-images.githubusercontent.com/405124/52805533-130f5180-3087-11e9-9bfd-50c1ee8637ac.png)

---

#### View of a dossier when a concurrent resolution is attempted
![attempt_at_concurrent_resolution](https://user-images.githubusercontent.com/405124/52805578-2e7a5c80-3087-11e9-9f0d-3d51074a2231.png)

---

#### Example logs for a concurrent attempt at resolving the same dossier

```
2019-02-14 18:24:41 INFO opengever.dossier [<Connection at 10a79c050>] Acquiring resolve lock for <DossierContainer at dossier-54>...
2019-02-14 18:24:41 INFO opengever.dossier [<Connection at 10a79c050>] Resolve lock acquired.
2019-02-14 18:24:52 INFO opengever.dossier [<Connection at 11ada90d0>] <DossierContainer at dossier-54> is resolve locked
2019-02-14 18:24:55 INFO opengever.dossier [<Connection at 11ada90d0>] <DossierContainer at dossier-54> is resolve locked via lock on <DossierContainer at /fd/ordnungssystem/bildung/dossier-54>
2019-02-14 18:24:55 INFO opengever.dossier [<Connection at 11ada90d0>] Concurrent attempt at resolving <DossierContainer at dossier-54> rejected
2019-02-14 18:24:55 INFO opengever.dossier [<Connection at 11ada90d0>] <DossierContainer at dossier-54> is resolve locked
2019-02-14 18:25:11 INFO opengever.dossier [<Connection at 10a79c050>] Successfully resolved <DossierContainer at dossier-54>
2019-02-14 18:25:11 INFO opengever.dossier [<Connection at 10a79c050>] Releasing resolve lock...
2019-02-14 18:25:11 INFO opengever.dossier [<Connection at 10a79c050>] Resolve lock released for <DossierContainer at dossier-54>
```
---

#### Example logs for an exception during resolution
```
2019-02-14 18:21:31 INFO opengever.dossier [<Connection at 10a79c050>] Acquiring resolve lock for <DossierContainer at dossier-54>...
2019-02-14 18:21:31 INFO opengever.dossier [<Connection at 10a79c050>] Resolve lock acquired.
2019-02-14 18:22:01 INFO opengever.dossier [<Connection at 10a79c050>] Releasing resolve lock...
2019-02-14 18:22:01 INFO opengever.dossier [<Connection at 10a79c050>] Resolve lock released for <DossierContainer at dossier-54>
2019-02-14 18:22:01 ERROR Zope.SiteErrorLog 1550164921.490.941219581584 http://localhost:8080/fd/ordnungssystem/bildung/dossier-54/transition-resolve
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module opengever.dossier.resolve, line 81, in __call__
  Module opengever.dossier.resolve, line 132, in execute_recursive_resolve
Exception: Boom
2019-02-14 18:22:01 ERROR Zope.SiteErrorLog 1550164921.50.650119714009 http://localhost:8080/fd/ordnungssystem/bildung/dossier-54/transition-resolve
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module opengever.dossier.resolve, line 81, in __call__
  Module opengever.dossier.resolve, line 132, in execute_recursive_resolve
Exception: Boom
```

*(**Note**: The exception being reported twice has nothing to do with the changes in this PR. I checked, the actual code is only run once, it's just a case of double-reporting that already happens in `maser`. The reason for this is that first ZPublisher [invokes`SiteErrorLog.raising()`](https://github.com/zopefoundation/Zope/blob/2.13.28/src/Zope2/App/startup.py#L214), later [looks up the exception handling view and calls it](https://github.com/zopefoundation/Zope/blob/2.13.28/src/Zope2/App/startup.py#L232), which finds our own error handling view, which then [calls `SiteErrorLog.raising()` again](https://github.com/4teamwork/opengever.core/blob/master/opengever/base/browser/errors.py#L48-L50). )*

deiferni commented on opengever/dossier/resolve.py:
```diff
@@ -59,44 +62,115 @@ def __call__(self, context):
+        api.portal.show_message(
+            message=_('Dossier is already being resolved'),
+            request=self.request, type='info')
+        return self.request.RESPONSE.redirect(self.context.absolute_url())
```
`self.redirect`?

lukasgraf replied:
That was the plan all along 😄 Thanks.

deiferni commented on opengever/dossier/resolve.py:
```diff
@@ -59,44 +62,115 @@ def __call__(self, context):
     def __call__(self):
+        # Ensure already resolved dossier can't be resolved again
+        if self.is_already_resolved():
+            self.show_already_resolved_msg()
```
below you redirect in the method that shows the message, here you don't. it is a detail but it helps the reader if both things work the same way.

lukasgraf replied:
Yep - I actually refactored all that stuff with the plan to homogenize the handling of redirects and showing messages. Somehow some stuff got left by the wayside.

deiferni commented on opengever/dossier/resolve.py:
```diff
@@ -59,44 +62,115 @@ def __call__(self, context):
+    def msg_being_resolved(self):
+        api.portal.show_message(
+            message=_('Dossier is already being resolved'),
+            request=self.request, type='info')
```
we could also show this as a warning-message 🤔.

lukasgraf replied:
I also considered this, and was on the fence for a while. In the end I went with `info` because there really isn't anything to warn the user about - we've all handled it for them. All that's left to do is for us to tell them nicely, and for them to reflect upon their actions 😉

njohner commented on opengever/dossier/resolve_lock.py:
```diff
@@ -0,0 +1,132 @@
+        age = datetime.now() - ts
+        expired = age > RESOLVE_LOCK_LIFETIME
+        if expired:
+            self.log("Resolve lock is expired (age: %s): %s" % (age, lockinfo))
```
Here expired locks could be cleared

lukasgraf replied:
@njohner I considered doing this. However, not here, because here we're only *testing* whether a lock is expired. Meaning, if we were to clean up locks here, we would also do this during reads, when just viewing a dossier triggers this check via the byline info. And writes on read are very bad for several reasons.

So I considered clearing existing locks during `acquire()` - but turns out, that really wouldn't add anything to what's being done right now. An existing lock simply gets overwritten with a fresh one during `acquire()`.

njohner replied:
Right, I still forget the to be careful with write on read...

njohner commented on opengever/dossier/resolve.py:
```diff
@@ -59,44 +62,115 @@ def __call__(self, context):
+            # In the case of a previous abort, the removal will be committed
+            # as as separate transaction, in the success case the removal will
+            # be part of the successful dossier resolution transaction.
+            resolve_lock.release(commit=True)
```
Ok that's good. Means that if we have a conflict during the transaction, we roll back to a state where the lock is still there, dossier not resolved and try again.
If the transaction fails 3 times, we will end up in a state with the dossier locked and not resolved. But that should not really happen anymore.

lukasgraf replied:
I believe the lock should even be removed in the case of conflicts and 3 retries - because then the `ConflictError` will actually be propagated, our `except` should catch it, and `finally` should still remove the lock. I think. I'll see if I can actually test this case, though it's quite tricky to reliably produce conflicts for testing this.

lukasgraf replied:
@njohner as discussed, you were right, this actually is an issue:

Because a `ConflictError` happens during `commit()`, and in my first approach I committed the removal of the lock together with all the work from resolving the dossiers, the transaction that conflicted was the same one that was supposed to remove the lock 😉 Which lead to a left over lock, which the retried request then stumbled upon (already locked).

So I changed it that we actually `commit()` at the end of the `try` clause (after successfully resolving). This means that a possible `ConflictError` would then manifest there and then, during that `commit()`, which can then by caughty by our `finally` like any other exception: The lock will be removed, removal committed, and we're ready for the retry of the request. Phew. 😉

njohner commented:
Missing resolves #5346

lukasgraf commented:
@njohner thanks for the review and the French translations! 💚

jone commented on opengever/dossier/resolve_lock.py:
```diff
@@ -0,0 +1,166 @@
+
+        item = self.context
+
+        while IDossierMarker.providedBy(item):
```
I prefer using `aq_chain` over a while loop since it is harder to produce an infinite loop when not using `while` 😉 
I'd do it like that:
```python
for item in filter(IDossierMarker.providedBy, aq_chain(item)):
    ...
```

I'm aware that the `while`  loop is not infinit now, but someone might "refactor" it and I practice the for / aq_chain approach as good habit 😜 

You could also make the function more DRY by checking `recursive` in the for-loop, I guess:

```python
for item in filter(IDossierMarker.providedBy, aq_chain(item)):
    ...
    if not recursive:
        return False
return False
```
Although this would have a different result when `context` is not providing `IDossierMarker`; not sure if this is even possible.

lukasgraf replied:
@jone as discussed: For now, I'm not going with the `filter(..., aq_chain)` approach yet. I would prefer this one too, but then I'd like to actually implement this in a well-tested helper function that is then used in all the places in `opengever.core` where we already do this. Well-tested would also include tests for the odd cases of multiple AQ wrappers, not just containment, that make the `aq_inner` necessary.

lukasgraf commented:
Updated:
- Improved the handling of ConflictErrors
- Made `ResolveLock.is_locked()` more DRY
- Worked in the French translations
- Made the use of `self.redirect` and early returns more consistent in the resolve view
- Fixed typo in commit message

Also tested again for MeetingDossiers.
